### ISSUE TITLE
[FR] Python 3.12 support

### ISSUE THREAD

**Issue by mandeep** [OpenBB-finance/OpenBB#6482]
Hi all,

Any plans to support Python 3.12?

I didn't see anything previously mentioned about this on the issue tracker.

**Comment by deeleeramone** [comment]
There are still a large number of packages that still do not support 3.12. For example, any library that relies on setuptools is not able to migrate. It's definitely on the, "would be nice to have", list; but it's not clear when it can be a reality because there are a lot of external factors.

**Comment by mdvthu** [comment]
On my installation (macOS), python3.12 (from brew) fails due to the pinned dependency on numpy 1.23.4. However, numpy 2.x.x installs fine in a python3.12 venv. Could openbb be made to work with numpy 2.0.0? I'm sure there would be other bits that would also need to be updated to remove the need for setuptools too but numpy 2.x.x is python3.12 compatible

**Comment by deeleeramone** [comment]
It's probably better to not use Homebrew for your Python environments because this will contaminate environment isolation with system PATH injection of Python libraries. This is especially evident when using `venv` and system installations of Python.

As for Python 3.12, Numpy is not the blocking element in the dependency chain.

If you are 'pinned' at 1.23.4, then you may be using version 3 of OpenBB.

The version of Numpy installed will depend on all the other extensions and components that are installed, but at minimum you should be with version 1.24.

**Comment by micahchenn** [comment]
I am also having the same issue. I have 

Collecting numpy==1.23.4 (from openbb)
  Using cached numpy-1.23.4.tar.gz (10.7 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [33 lines of output]
      Traceback (most recent call last):
        File "C:\Users\micah\Desktop\Harbinger_Backend\env\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
        File "C:\Users\micah\Desktop\Harbinger_Backend\env\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\micah\Desktop\Harbinger_Backend\env\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 112, in get_requires_for_build_wheel
          backend = _build_backend()
                    ^^^^^^^^^^^^^^^^
        File "C:\Users\micah\Desktop\Harbinger_Backend\env\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 77, in _build_backend
          obj = import_module(mod_path)
                ^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\micah\AppData\Local\Programs\Python\Python312\Lib\importlib\__init__.py", line 90, in import_module
          return _bootstrap._gcd_import(name[level:], package, level)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "<frozen importlib._bootstrap>", line 1381, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1354, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1304, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "<frozen importlib._bootstrap>", line 1381, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1354, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1325, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 929, in _load_unlocked
        File "<frozen importlib._bootstrap_external>", line 994, in exec_module
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "C:\Users\micah\AppData\Local\Temp\pip-build-env-h5p5euqx\overlay\Lib\site-packages\setuptools\__init__.py", line 16, in <module>
          import setuptools.version
        File "C:\Users\micah\AppData\Local\Temp\pip-build-env-h5p5euqx\overlay\Lib\site-packages\setuptools\version.py", line 1, in <module>
          import pkg_resources
        File "C:\Users\micah\AppData\Local\Temp\pip-build-env-h5p5euqx\overlay\Lib\site-packages\pkg_resources\__init__.py", line 2172, in <module>
          register_finder(pkgutil.ImpImporter, find_on_path)
                          ^^^^^^^^^^^^^^^^^^^
      AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.



I have numpy installed ── numpy [required: >=1.16.5, installed: 1.26.4]


However, I am unable to download openBB due to this error.

How can i fix this?

**Comment by deeleeramone** [comment]
This is resolved with #6572

**Comment by alexchandel** [comment]
A similar error exists on 3.13.  Any way we can update these ancient dependencies to the latest?

```txt
Collecting llama-index<0.9.0,>=0.8.8 (from openbb[all])
  Using cached llama_index-0.8.42-py3-none-any.whl.metadata (5.0 kB)
Collecting matplotlib<4.0.0,>=3.3.3 (from openbb[all])
  Using cached matplotlib-3.10.0-cp313-cp313-win_amd64.whl.metadata (11 kB)
Collecting mplfinance<0.13.0,>=0.12.7-alpha.7 (from openbb[all])
  Using cached mplfinance-0.12.10b0-py3-none-any.whl.metadata (19 kB)
Collecting mstarpy<0.0.5,>=0.0.4 (from openbb[all])
  Using cached mstarpy-0.0.4-py3-none-any.whl.metadata (5.3 kB)
Collecting nixtlats<0.2.0,>=0.1.11 (from openbb[all])
  Using cached nixtlats-0.1.21-py3-none-any.whl.metadata (4.5 kB)
Collecting nltk<4.0.0,>=3.8.1 (from openbb[all])
  Using cached nltk-3.9.1-py3-none-any.whl.metadata (2.9 kB)
Collecting numpy==1.23.4 (from openbb[all])
  Using cached numpy-1.23.4.tar.gz (10.7 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  error: subprocess-exited-with-error

  × Getting requirements to build wheel did not run successfully.
  │ exit code: 1
  ╰─> [32 lines of output]
      Traceback (most recent call last):
        File "C:\Users\me\Projects\OpenBB\.venv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 353, in <module>
          main()
          ~~~~^^
        File "C:\Users\me\Projects\OpenBB\.venv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 335, in main
          json_out['return_val'] = hook(**hook_input['kwargs'])
                                   ~~~~^^^^^^^^^^^^^^^^^^^^^^^^
        File "C:\Users\me\Projects\OpenBB\.venv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 112, in get_requires_for_build_wheel
          backend = _build_backend()
        File "C:\Users\me\Projects\OpenBB\.venv\Lib\site-packages\pip\_vendor\pyproject_hooks\_in_process\_in_process.py", line 77, in _build_backend
          obj = import_module(mod_path)
        File "c:\Program Files\Python313\Lib\importlib\__init__.py", line 88, in import_module
          return _bootstrap._gcd_import(name[level:], package, level)
                 ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1310, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
        File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
        File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
        File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
        File "<frozen importlib._bootstrap_external>", line 1026, in exec_module
        File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
        File "C:\Users\me\AppData\Local\Temp\pip-build-env-ly9mfm8z\overlay\Lib\site-packages\setuptools\__init__.py", line 16, in <module>
          import setuptools.version
        File "C:\Users\me\AppData\Local\Temp\pip-build-env-ly9mfm8z\overlay\Lib\site-packages\setuptools\version.py", line 1, in <module>
          import pkg_resources
        File "C:\Users\me\AppData\Local\Temp\pip-build-env-ly9mfm8z\overlay\Lib\site-packages\pkg_resources\__init__.py", line 2172, in <module>
          register_finder(pkgutil.ImpImporter, find_on_path)
                          ^^^^^^^^^^^^^^^^^^^
      AttributeError: module 'pkgutil' has no attribute 'ImpImporter'. Did you mean: 'zipimporter'?
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build wheel did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
```
